lib/Sema/CSApply.cpp

```diff
@@ -6303,6 +6303,10 @@ void ExprRewriter::peepholeArrayUpcast(ArrayExpr *expr, Type toType,
                                        ConstraintLocatorBuilder locator) {
   // Update the type of the array literal.
   cs.setType(expr, toType);
+  // FIXME: finish{Array,Dictionary}Expr invokes cacheExprTypes after forming
+  // the semantic expression for the dictionary literal, which will undo the
+  // type we set here if this dictionary literal is nested.
+  expr->setType(nullptr);
```

Comment:
Maybe a better fix for this situation would be for `CacheExprTypes` not to override types which are already in the cache?

Reply:
I could try that. Are you OK with this as a minimal-impact fix for the 4.2 branch?

Reply:
Yeah, that sounds good!